### ISSUE TITLE
Using javascript libaries with Vugu

### ISSUE THREAD

**Issue by owenwaller** [vugu/vugu#230]
**Is your feature request related to a problem? Please describe.**

We are considering using Vugu to write our UI in, as the project is entirely Go based at the minute.

A quick trial shows that we can do most of what we need - basic tabular data, buttons and forms. It also shows that we have the potential for a lot of code reuse in the browser. The process has been both quick and relatively pain free for the simple trials we've done so far.

But, our pain point is visualization. We have a lot of charts to generate, and those charts need to be interactive so the rendering has to be done in the browser.

We are looking at using a Vega + d3.js combination ATM.
https://vega.github.io/vega/
https://d3js.org/

Vega essentially let's you describe a chart as json, and then uses d3.js to do the rendering.

**Describe the solution you'd like**
What we would like to do is:

Have a vugu file that references the necessary javascript and css files. 
In the same file, we need to add some inline <script> tags containing javascript to the vugu file. This is necessary to define a view in Vega which controls the interactivity of the chart.
Have a go object, that makes the http requests to the backend to generate the JSON we need to pass to the Vega parser.
From Go we need to create a javascript variable we require, and assign that the value of the JSON the server sent us. This is how we have to pass the JSON to the Vega parser.
Then on the next render pass, now that the data is available, have the Javascript <script> tags execute to produce the chart, including any javascript events its defined (e.g. mousein, mouseout)

The problems seem to be:

Vugu won't allow a <script> tag (as the docs state) in a .vugu file
Setting a javascript variable from Go - I'm assuming sys/js will help here
Having javascript in the <script> block execute
Having any events defined in the JSON execute, which Vega will map to Javascript events execute
Working out the file structure of the root.vugi file (and what it needs to contain) + the other components.

**Describe alternatives you've considered**

We can generate the charts server side, as SVG's. But that looses the interactivity which we need.
See; https://github.com/gravwell/govega

If this isn't possible with Vugu, then our only other choice is to look at a wholly Typescript based UI.

That means we loose any code reuse we have on the client side, which is going to be painful. That's x2 the work in a second language (as we will always have go client side libraries and cmd line tools for each server, for both integration (go -> go) and testing) and twice the testing (which we now also need for the new javascript client side libraries).

Just to be clear from a users point of view, the chart and all of the associated data need to be on the same "webpage"., which is a page we would like to be generated with Vugu. So we can't mix the languages to build the UI.

**Additional context**

For a indication of what we are trying to do look at the bar chart example in Vega.
https://vega.github.io/vega/examples/bar-chart/

if you look at the source code of that page, we basically need to Vugu to support whats between the <script> blocks + any javascript libraries in the page header. See the use of the `spec` variable in this code snippit, from source code of the Vega bar chart example page:
```
var spec = {...}

var view = new vega.View(vega.parse(spec), {
  loader: vega.loader({baseURL: '/vega/'}),
  logLevel: vega.Warn,
  renderer: 'svg'
}).initialize('#bar-chart').hover().run();

```

**Comment by bradleypeabody** [comment]
Thanks for writing this up and I agree it would be excellent to get this aspect fixed up in Vugu.  To response to your points:

> Vugu won't allow a <script> tag (as the docs state) in a .vugu file
Yes.  This is fixable, and it's not necessarily a huge task, it just requires carefully adding the necessary processing at several points in the flow, as well as working out the "lifecycle" of such a script tag when Vugu emits it in the document. Possibly it can be done the same way as CSS where it gets emitted when indicated by the currently active components and then removed when not.  But unlike CSS which is declarative, JS is executed and can have side effects, so I'm not sure if re-executing JS whenever it is added to a page is the most useful behavior (maybe it is, needs more consideration).

> Setting a javascript variable from Go - I'm assuming sys/js will help here
This should be simple.  If the stuff above is figured out, code like this can easily be included in an example.

> Having javascript in the <script> block execute
Covered above

> Having any events defined in the JSON execute, which Vega will map to Javascript events execute
Possibly we need some more explicit way to call from JS back into Go.  The Go runtime provides some facility for this, but I think some working examples would need to be figured out and then we'd see if it's too clunky and needs additional tooling.

> Working out the file structure of the root.vugi file (and what it needs to contain) + the other components.
Yeah I've been meaning to put out a new example with a recommended file structure but haven't gotten to it yet.  I'll see if I can get this done in the coming days.

**Comment by owenwaller** [comment]
Hi Bradley,

Many thanks for the reply.

A little more research on my part has suggested there might be a way.

One technique suggested in React based apps is to get React to dynamically generate and insert the `<script>` element. See:
https://stackoverflow.com/questions/34424845/adding-script-tag-to-react-jsx
for example.

Now with Vugu + /syscall/js I wonder is something similar is possible? 

With this approach the `<script>` tag would be created in Go, so Vugu might not need any changes. The script itself can then be set via the tags `src` attribute, or defined as a child element, and attached to the parent script tag.

I haven't tried this yet, but it feels like it has a fighting chance of working. I need to try an experiment over the next few days. If this works there that can perhaps be wrapped up into a useful struct, for reuse. Does Vugu have a package that makes the DOM manipulation this would require easier?

In our case the "lifetime" of the script if the lifetime of the visibility of the chart, or whatever div the chart is in. But that may not generally be the case. So I agree with you the lifetime of a script needs to be considered. 

What is the situation in Javascript normally? I had always assumed that any functions/events in a  `<script>` was available until the page changed i.e. new URL or they where explicitly removed from the DOM.

The JS events might be more tricky. Essentially the Vega JS code will create the event function callbacks for us, and attach them to the JS events - i.e mousein/mouseout - and then attach these to the DOM elements the chart uses.

So I was hoping that the Javascript interpreter in the browser would just do the correct thing in this case - and even is an event after all regardless of where it was defined i.e.  in Javascript or as WASM (from Go).

 Owen

**Comment by bradleypeabody** [comment]
What you're describing can definitely work and I've done similar things before.  Here's a working example of dynamically inserting (and thus executing) JS on a page:

https://play.vugu.org/?p=H4sIAAAAAAAC_5STwW7bMBBEz-ZXbPcisnCpnFvZaJoEAQo0LZp8QBhyIzOlSEFaGTaC_ntByU7tOIf2JAmaNzsjrirn10shZtXDwJwifLbB218LtPrcuVvb-ZYlrSmywuUd9VyVk3ApRFVObNWPMuBtSws0bRu8NexTLDcf6oRLIXzTpo4BQ6pRiKyDnykx9NwNluFZ_BbicYgWpIX3-Y2CV9NhPdSDvvz-7WrMAs9iFlKtf3Q-8qPEy200jbcmhC3Yjgz7WEOfGoKvt6iEmJUlrE0HKzIOFuCSHRqKrGviq0D5tv-yvTP1jWlIFllVKO2ZGnmmPonZiH1cwFOvr0N6MEEqfU0scW-ESl-YECS-ZYhzwOzwIsrGOIcz9ZJr9wUPko0taOcli0lQ5DA77T_FOXLJOSYYp8nTg-6Jz5k7_zAwySKfTjGHgmnD5ZNZm5PRO-9DLFtncLweg_i3pU2OTjve0YZvkiNZmEAdS1xRCOkdqkKJ2Yj8R9W9Gc7h_pXd_VFp07YU3cXKByfzkJN6BwKcj9EnPp_kET1R-zV5k95JxPHSmui01uBSJFT5F6jKSbj8EwAA__-q4HfVlQMAAA%3D%3D

> What is the situation in Javascript normally? I had always assumed that any functions/events in a <script> was available until the page changed i.e. new URL or they where explicitly removed from the DOM.

Script tags are executed as soon as they are loaded.  For inline scripts this means immediately.  For scripts which pull from a remote URL this means once the script is actually available in the browser after downloading it. For a JS function, the "execution" means, yes, declaring the function so it is available to be called at any time later.  Removing the script tag does not (I'm like 99% sure) have any affect on functions that have already been declared.  But you see it makes a big different what you put in the script tag.  Functions just get declared, but statements get executed.

For a typical web page a script tag is either included in the page at startup or dynamically created once and executes and that's it.  Any further code that needs to run is hooked up via events or other callbacks.  The situation is a bit different in Vugu because of the semantic meaning of a Vugu component - it is only supposed to "be present" when the component is rendered and also might have a number of instances of the same component in different places in the page.  And this is different from Vue and React where the JS associated with a component has very specific rules about how it interacts with that exact component.

So I think the key thing to implement JS script tags is Vugu is deciding what behavior is both A) realistic to implement and B) useful to developers.  The main options on the table would seem to be:

* Make <script> tags execute exactly once, the first time they are "rendered".
* Make <script> tags execute each time they are "rendered".

Each of these has tradeoffs involved - if we run the JS only once then I think developers are likely to be confused about the lifecycle involved.  This means that for example if there is a script tag that appears in a loop it will only be executed the first iteration.  Could work, but certainly not obvious.  Conversely if we execute every time I fear it could encourage people to do very inefficient things, like put huge blocks of code in components that only need to be declared once and then called but instead the entire block would be redeclared constantly.

Anyway, you have the workaround above, but in terms of the Vugu feature the main barrier is arriving at an idea of what is a good behavior that is useful enough to a wide swath of developers and won't encourage misuse.

**Comment by owenwaller** [comment]
HI,

Apoligises for the delay in replying. Thanks to @hirohido for reminding me to reply. Hopefully this will help you.

@bradleypeabody The example was just enough to work this out. Many thanks for that.

Based on the example I've wrapped what I needed into two simple functions:

```
func (c *Root) AddScript(src string) {
	head := js.Global().Get("document").Call("getElementsByTagName", "head").Call("item", 0)
	script := js.Global().Get("document").Call("createElement", "script")
	script.Call("setAttribute", "type", "text/javascript")
        // rather than create an inner child element that contains the javascript we want to execute
        // we can use the "src" property of the script tag, to set where we want the script to be downloaded from.
        script.Call("setAttribute", "src", src)
	head.Call("appendChild", script)
}

func (c *Root) AddStyleSheet(src string) {
	head := js.Global().Get("document").Call("getElementsByTagName", "head").Call("item", 0)
	link := js.Global().Get("document").Call("createElement", "link")
	link.Call("setAttribute", "rel", "stylesheet")
        link.Call("setAttribute", "href", src)
	head.Call("appendChild", link)
}
```

These should both add a script and a style sheet into the head element of the DOM. If my understand of things is correct then when the DOM is re-evaluated both the javascript and the style sheet should be available. The javascript functions are loaded at this point but not executed. (Global) javascript variables that are defined in the script would be evaluated/executed at this point.

Once I saw the example, I realized that I didn't need an inline script block, as I had originally thought. With a little refactoring of the javascript function I wanted to call in the browser, I could just serve the javascript I needed directly from the server.

Once I had this, I could then call the javascript function I needed via the `syscall/js` package as normal i.e.

```
 js.Global().Call("my-function", blob)
```

Where the `blob` is a java script object that I need to pass to `my-function`.

The javascript code that I am loading will bind javascript functions to events - think hover events - that I have no direct control over. I had been concerned that this may not work with Vugu. The good news is, that it does work and is completely transparent.

Note: None of these events are bound back to a Vugu component. This is all pure javascript at this point.

In terms of Vugu, does this now open up a slightly different approach?

Rather than a new vugu tag, which is what I originally thought might be required, why don't we add a small set of functions, similar to the above?

These new functions would always inject into the head element of the DOM, so normal javascript scoping/lifecycle rules apply.

I am thinking off something like:

```
func AddScriptToHead(srcUrl string) {....}
func AddStyleSheetToHead(srcUrl string) {...}
func AddScriptToHeadWithAttributes(srcUrl string, attrs ScriptAttributes) {....}
func AddStyleSheetToHeadWithAttributes(srcUrl string, attrs StyleSheetAttributes) {....}
...
... etc e.g. for the meta, author tags, et all ...
```

We don't provide a "Remove" function to remove these from the DOM, as this seems to be a be a no op as @bradleypeabody  suggests.

In terms of adding a scripts inline, rather than provide this, would it be better if Vugu documented how to do it - based on the example - along with a warning about the javascript lifecycle, not matching the component lifecycle as you might expect. Would that be enough to cover the common use cases?

This also kicks the issue of the javascript lifecycle vs the component lifecycle down the road until it's been worked though?

Is this a workable solution?

@hirohido My use case was pretty specific to the javascript charting library I am trying to use. What is your use case and how does it differ?

@bradleypeabody Now I've got this to work there is an obvious alternative solution. Since I'm now injecting the javascript and the stylesheet directly into the head element of the DOM, I could have just put these in a root..vugu component and used full HTML mode. I haven't tried this yet, as I'll need to rework my root.vugu file.

**Comment by owenwaller** [comment]
As I started this issue....

I think this is probably worth adding the new functions to insert a javascript or css into the head element of a page in sort fort is utility or helper package. I can't be the only one that needed this functionality.

**Comment by powerman** [comment]
> I can't be the only one that needed this functionality.

You're not the only one, for sure. But the actual need is to have a (recommended) way(s) either how to use Vugu with **any** existing JS libs/frameworks or how to (re)implement features of existing JS libs/frameworks in Go using just Vugu. Because at the end we all need modern dynamic app working in browser, with all sorts of bells and whistles.

**Comment by owenwaller** [comment]
@powerman what I am proposing would load any Javascript script into a page. 

How you call the Javascript functions that are defined within the script is a separate issue.

Once the JS script is loaded into the browser then you use the [Go sys/js](https://pkg.go.dev/syscall/js) package from your go files, or from within your `<script type="application/x-go">` blocks, if that is what your using, to call the javascript functions.

That's what we did and it worked just fine. I can't release the source code, but we can possibly add some sort of example in the vugu documentation if that would help.

If you want a Go/WASM wrapper around a specific library, then that IMHO is not a job for Vugu, but for you as a library user or the library authors. 

To do this you'll need to create a binding from Go to JS. Looking at of of these will show you the sort of work you need to do with the Go sys/js package to get things to work. There are some examples on the [Gopher JS Bindings](https://github.com/gopherjs/gopherjs/wiki/bindings) page. Which is based on the same Go sys/js package as Vugu uses. 

I haven't tried any of these as that is not what the project I am working with needed so YMMV.

@bradleypeabody AFAIK it's still the case that we can't execute raw JS inside a script tag.
